### PR TITLE
[+] Authentication - Add Google authentication.

### PR DESCRIPTION
**Pull request by slimee**

### CODE REVIEW

**Comment by slimee** [comment]
An important automated test is missing for google authentication.
Still looking for a way to make it work.

File: src/services/authenticator.js [hunk 1]
```diff
@@ -25,26 +27,27 @@ function Authenticator() {
...
-        email: config.email,
-        password: config.password,
-      })
-      .then((response) => response.body)
-      .then((auth) => {
-        config.authToken = auth.token;
-        return fs.writeFileSync(`${os.homedir()}/.forestrc`, auth.token);
-      });
+  this.validateToken = (token) => !!this.verify(token) || 'Invalid token. Please enter your'
+    + ' authentication token.';
```
Comment (rap2hpoutre) [inline src/services/authenticator.js]: This two lines could be refactored to be more readable (just a suggestion though)

File: src/utils/messages.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  ERROR_UNEXPECTED: 'An unexpected error occurred. Please reach out for help in our Slack community or create a Github issue with following error:',
+  ERROR_MISSING_PROJECT_NAME: 'Missing project name in the command.',
+  HINT_MISSING_PROJECT_NAME: 'Please specify a project name. Type lumber help for more information.',
+  ERROR_NOT_PARSABLE_CONNECTION_URL: 'Cannot parse the database dialect. Please, check the syntax of the database connection string.',
+  HINT_DIRECTORY_ALREADY_EXISTS: 'Please retry with another project name.',
```
Comment (rap2hpoutre) [inline src/utils/messages.js]: These lines are only for lumber.

File: src/services/authenticator.js [hunk 3]
```diff
@@ -71,6 +74,77 @@ function Authenticator() {
...
+    }]);
+    return sessionToken;
+  };
+
+  this.loginWithPassword = async (email, password) => {
+    if (!password) {
+      ({ password } = await inquirer.prompt([{
+        type: 'password',
+        name: 'password',
+        message: 'What\'s your Forest Admin password:',
```
Comment (rap2hpoutre) [inline src/services/authenticator.js]: Just a suggestion

```suggestion
        message: 'What is your Forest Admin password:',
```

Or (to be consistant with token prompt):

```suggestion
        message: 'Enter your Forest Admin password:',
```

File: src/services/authenticator.js [hunk 4]
```diff
@@ -71,6 +74,77 @@ function Authenticator() {
...
+    }
+
+    return this.loginWithPassword(email, password);
+  };
+
+  this.promptEmail = async () => {
+    const { email } = await inquirer.prompt([{
+      type: 'input',
+      name: 'email',
+      message: 'What\'s your email address?',
```
Comment (rap2hpoutre) [inline src/services/authenticator.js]: Just a suggestion

```suggestion
      message: 'What is your email address?',
```

File: src/utils/regexs.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  EMAIL_REGEX: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+  // NOTICE: The forest password should be composed of digit, at least on capital letter
+  //         and one lower case letter. It also accepts special characters except whitespaces.
+  PASSWORD_REGEX: /^(?=\S*?[A-Z])(?=\S*?[a-z])((?=\S*?[0-9]))\S{8,}$/,
```
Comment (rap2hpoutre) [inline src/utils/regexs.js]: This is not in use in Toolbelt.

File: test/commands/login.test.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  });
+
+  describe('with trivial mail and valid token in args', () => {
+    fancy
+      .stdout()
+      .stderr()
+      .command(['login', '-e', 'smile@gmail.com', '-t', jwt.sign({}, 'key', { expiresIn: '1day' })])
+      .it('should warn about bad token', (ctx) => {
+        expect(ctx.stdout).to.contain('Login successful');
+      });
```
Comment (rap2hpoutre) [inline test/commands/login.test.js]: It seems it does not "warn about bad token".

File: src/commands/login.js [hunk 7]
```diff
@@ -1,32 +1,42 @@
...
+  async run() {
     await authenticator
       .logout({ log: false });
 
-    const config = await Prompter([
-      'email',
-      'password',
-    ]);
-
+    const { flags: config } = this.parse(LoginCommand);
```
[the remark this answers is not in the document]
Comment (slimee) [reply src/commands/login.js]: 🎩

File: src/services/authenticator.js [hunk 8]
```diff
@@ -71,6 +74,77 @@ function Authenticator() {
...
+
+  this.promptEmail = async () => {
+    const { email } = await inquirer.prompt([{
+      type: 'input',
+      name: 'email',
+      message: 'What is your email address?',
+      validate: (input) => {
+        if (!Joi.string().email().validate(input).error) {
+          return true;
+        }
```
Comment (rap2hpoutre) [inline src/services/authenticator.js]: We only validate email when prompted, not when sent via command line. We could consider validate it in all cases:

Capture d’écran 2020-03-12 à 14 35 38

Comment (slimee) [reply src/services/authenticator.js]: Let me fix this

**Pull request merged**